[docs/data] Batch training with Ray Dataset example (SSML)

christy opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

This PR adds a notebook showcasing how to conduct batch training with Ray Dataset, mirroring the Ray Core example.

## Related issue number

<!-- For example: "Closes #1234" -->
Related PR https://github.com/ray-project/ray/pull/28863

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jjyao commented:
If it's batch training with ray datasets then it should be put under data instead of ray-core? Ray-core already has a batch training version that doesn't use datasets. cc @Yard1

Yard1 commented:
Yes, I am working on getting this PR over the finish line as Christy is on PTO - already made this change locally :)

clarkzinzow commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -0,0 +1,1228 @@
+   "source": [
+    "# A pandas DataFrame UDF for transforming the Dataset in parallel.\n",
+    "def transform_batch(df: pd.DataFrame) -> pd.DataFrame:\n",
+    "    df[\"trip_duration\"] = (df[\"dropoff_at\"] - df[\"pickup_at\"]).dt.seconds\n",
```
Nit: Use the `.loc[]` API to mutate the DataFrame rather than create a new copy.

```suggestion
    "    df.loc[\"trip_duration\"] = (df[\"dropoff_at\"] - df[\"pickup_at\"]).dt.seconds\n",
```

clarkzinzow commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -0,0 +1,1228 @@
+    "    df[\"trip_duration\"] = (df[\"dropoff_at\"] - df[\"pickup_at\"]).dt.seconds\n",
+    "    df = df[df[\"trip_duration\"] > 60]\n",
+    "    df.drop([\"dropoff_at\", \"pickup_at\", \"pickup_location_id\"], axis=1, inplace=True)\n",
+    "    df[\"dropoff_location_id\"] = df[\"dropoff_location_id\"].fillna(-1)\n",
```
Nit: Same here.
```suggestion
    "    df.loc[\"dropoff_location_id\"] = df[\"dropoff_location_id\"].fillna(-1)\n",
```

clarkzinzow commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -0,0 +1,1228 @@
+    "\n",
+    "# A Pandas DataFrame UDF for transforming the underlying blocks of a Dataset in parallel.\n",
+    "def transform_batch(df: pd.DataFrame) -> pd.DataFrame:\n",
+    "    df[\"trip_duration\"] = (df[\"dropoff_at\"] - df[\"pickup_at\"]).dt.seconds\n",
```
Nit: Same here.
```suggestion
    "    df.loc[\"trip_duration\"] = (df[\"dropoff_at\"] - df[\"pickup_at\"]).dt.seconds\n",
```

clarkzinzow commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -0,0 +1,1228 @@
+    "    df[\"trip_duration\"] = (df[\"dropoff_at\"] - df[\"pickup_at\"]).dt.seconds\n",
+    "    df = df[df[\"trip_duration\"] > 60]\n",
+    "    df.drop([\"dropoff_at\", \"pickup_at\", \"pickup_location_id\"], axis=1, inplace=True)\n",
+    "    df[\"dropoff_location_id\"] = df[\"dropoff_location_id\"].fillna(-1)\n",
```
Nit: Same here.
```suggestion
    "    df.loc[\"dropoff_location_id\"] = df[\"dropoff_location_id\"].fillna(-1)\n",
```

clarkzinzow commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -0,0 +1,1228 @@
+   ],
+   "source": [
+    "# sort values by location id\n",
+    "results_df = results.to_pandas(limit=float(\"inf\"))\n",
```
Nit: The `.to_pandas()` API has a default limit of 100k, which should be more than enough after the `.map_groups()` operation.
```suggestion
    "results_df = results.to_pandas()\n",
```

clarkzinzow reviewed (commented):
LGTM overall in terms of Datasets implementation, mostly nits! I'll give the content a closer, more holistic look later today.

christy commented:
Updated polish for smoke test notebook doc/source/data/examples/batch_training.ipynb
* Added TOC with links within the notebook
* Shortened the intro Ray Datasets part
* Added multiple models
* Tried the .loc comments from Clark, but still got pandas warning messages about modifying a slice.  Cleanest for me was to copy the df inside the function, modify the copy, return the copy.

Yard1 commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -0,0 +1,994 @@
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# # The following reads everything in the files into a ray dataset.\n",
+    "# ds = ray.data.read_parquet(s3_files)\n",
+    "# ds"
```
should this be commented out? if so, can we say why we are not running it right now?

christy replied:
I'll add a comment to the notebook and push again.

christy commented:
Another update polish for smoke test notebook doc/source/data/examples/batch_training.ipynb
- Completely separate out data processing before data groupby train
- Repartition into 14 batches instead of 16 for benchmarking on 16 CPU

Yard1 commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -585,9 +588,11 @@
     "    \n",
     "    # Assemble return as a pandas dataframe.\n",
-    "    return_df = pd.DataFrame({'model': [model], 'error': [error]})\n",
+    "    return_df = pd.DataFrame({'model': [str(model)], 'error': [error]})\n",
```
why are we returning a string? the model would be better imo, as it's fitted and ready to be used/inspected

christy replied:
I was worried about space.  I guess these sklearn binary models are small enough.  Will make the change, thanks.

Yard1 commented on doc/source/data/examples/batch_training.ipynb:
```diff
@@ -683,14 +689,15 @@
     "        # Transform the input pandas AND fit_and_evaluate the transformed pandas\n",
-    "        transformed_df = transform_batch(df)\n",
-    "        results_df = train_and_evaluate(transformed_df, MODELS, location_id)\n",
+    "        # transformed_df = transform_batch(df)\n",
+    "        results_df = train_and_evaluate(df, MODELS, location_id)\n",
```
remove?
